PR: https://github.com/odoo/odoo/pull/203987

From: 61b5d9d78ca21ce7adcbe0f88e63ec9a5a4a3f89
From: clesgow
Date: 2025-08-29 06:50:51

Breaking data model changes scores: del:11 + add:1, change matches:
+    quant_ids = fields.One2many('stock.quant', 'package_id', 'Bulk Content', readonly=True,
-    name = fields.Char(
-    quant_ids = fields.One2many('stock.quant', 'package_id', 'Bulk Content', readonly=True,
-    package_type_id = fields.Many2one(
-    location_id = fields.Many2one(
-    company_id = fields.Many2one(
-    owner_id = fields.Many2one(
-    package_use = fields.Selection([
-    shipping_weight = fields.Float(string='Shipping Weight', help="Total weight of the package.")
-    valid_sscc = fields.Boolean('Package name is valid SSCC', compute='_compute_valid_sscc')
-    pack_date = fields.Date('Pack Date', default=fields.Date.today)

Total Changes: 342

[REF] stock, *: rename stock.quant.package

Rename the package model from `stock.quant.package` to `stock.package`
to better match the stock naming scheme.

Also moves the code relative to `stock.package` to their own files for
easier searches.

Task-4314600

Part-of: odoo/odoo#203987
Related: odoo/enterprise#82453
Related: odoo/upgrade#7576
Signed-off-by: Arnold Moyaux (arm) <arm@odoo.com>

================================= pseudo patch: =================================

--- a/addons/stock/models/__init__.py
+++ b/addons/stock/models/__init__.py
@@ -23,4 +23,5 @@ from . import product
 from . import product_catalog_mixin
 from . import stock_package_level
 from . import stock_package_type
+from . import stock_package
 from . import stock_storage_category

--- a/addons/stock/models/stock_move.py
+++ b/addons/stock/models/stock_move.py
@@ -1712,7 +1712,7 @@ Please change the quantity done or the rounding precision in your settings.""",
         if not lot_id:
             lot_id = self.env['stock.lot']
         if not package_id:
-            package_id = self.env['stock.quant.package']
+            package_id = self.env['stock.package']
         if not owner_id:
             owner_id = self.env['res.partner']
 

--- a/addons/stock/models/stock_move_line.py
+++ b/addons/stock/models/stock_move_line.py
@@ -41,7 +41,7 @@ class StockMoveLine(models.Model):
         copy=False, compute='_compute_quantity_product_uom', store=True)
     picked = fields.Boolean('Picked', compute='_compute_picked', store=True, readonly=False, copy=False)
     package_id = fields.Many2one(
-        'stock.quant.package', 'Source Package', ondelete='restrict',
+        'stock.package', 'Source Package', ondelete='restrict',
         check_company=True,
         domain="[('location_id', '=', location_id)]")
     package_level_id = fields.Many2one('stock.package_level', 'Package Level', check_company=True, index='btree_not_null')
@@ -50,7 +50,7 @@ class StockMoveLine(models.Model):
         domain="[('product_id', '=', product_id)]", check_company=True)
     lot_name = fields.Char('Lot/Serial Number Name')
     result_package_id = fields.Many2one(
-        'stock.quant.package', 'Destination Package',
+        'stock.package', 'Destination Package',
         ondelete='restrict', required=False, check_company=True,
         domain="['|', '|', ('location_id', '=', False), ('location_id', '=', location_dest_id), ('id', '=', package_id)]",
         help="If set, the operations are packed into this package")
@@ -430,8 +430,8 @@ class StockMoveLine(models.Model):
             ('location_id', 'stock.location'),
             ('location_dest_id', 'stock.location'),
             ('lot_id', 'stock.lot'),
-            ('package_id', 'stock.quant.package'),
-            ('result_package_id', 'stock.quant.package'),
+            ('package_id', 'stock.package'),
+            ('result_package_id', 'stock.package'),
             ('owner_id', 'res.partner'),
             ('product_uom_id', 'uom.uom')
         ]
@@ -758,9 +758,9 @@ class StockMoveLine(models.Model):
         if 'location_dest_id' in vals:
             data['location_dest_name'] = self.env['stock.location'].browse(vals.get('location_dest_id')).name
         if 'package_id' in vals and vals['package_id'] != move.package_id.id:
-            data['package_name'] = self.env['stock.quant.package'].browse(vals.get('package_id')).name
+            data['package_name'] = self.env['stock.package'].browse(vals.get('package_id')).name
         if 'package_result_id' in vals and vals['package_result_id'] != move.package_result_id.id:
-            data['result_package_name'] = self.env['stock.quant.package'].browse(vals.get('result_package_id')).name
+            data['result_package_name'] = self.env['stock.package'].browse(vals.get('result_package_id')).name
         if 'owner_id' in vals and vals['owner_id'] != move.owner_id.id:
             data['owner_name'] = self.env['res.partner'].browse(vals.get('owner_id')).name
         record.message_post_with_source(
@@ -1010,7 +1010,7 @@ class StockMoveLine(models.Model):
             }
 
     def _put_in_pack(self):
-        package = self.env['stock.quant.package'].create({})
+        package = self.env['stock.package'].create({})
         package_type = self.move_id.packaging_uom_id.package_type_id
         if len(package_type) == 1:
             package.package_type_id = package_type
@@ -1036,7 +1036,7 @@ class StockMoveLine(models.Model):
     def _post_put_in_pack_hook(self, package, **kwargs):
         if package and self.picking_type_id.auto_print_package_label:
             if self.picking_type_id.package_label_to_print == 'pdf':
-                action = self.env.ref("stock.action_report_quant_package_barcode_small").report_action(package.id, config=False)
+                action = self.env.ref("stock.action_report_package_barcode_small").report_action(package.id, config=False)
             elif self.picking_type_id.package_label_to_print == 'zpl':
                 action = self.env.ref("stock.label_package_template").report_action(package.id, config=False)
             if action:

--- a/None
+++ b/addons/stock/models/stock_package.py
@@ -0,0 +1,158 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from collections import defaultdict
+
+from odoo import api, fields, models
+from odoo.exceptions import UserError
+from odoo.fields import Domain
+from odoo.tools import groupby
+from odoo.tools.barcode import check_barcode_encoding
+from odoo.tools.float_utils import float_is_zero
+
+
+class StockPackage(models.Model):
+    """ Packages containing quants and/or other packages """
+    _name = 'stock.package'
+    _description = "Packages"
+    _order = 'name'
+
+    name = fields.Char(
+        'Package Reference', copy=False, index='trigram', required=True,
+        default=lambda self: self.env['ir.sequence'].next_by_code('stock.package') or self.env._('Unknown Pack'))
+    quant_ids = fields.One2many('stock.quant', 'package_id', 'Bulk Content', readonly=True,
+        domain=['|', ('quantity', '!=', 0), ('reserved_quantity', '!=', 0)])
+    package_type_id = fields.Many2one(
+        'stock.package.type', 'Package Type', index=True)
+    location_id = fields.Many2one(
+        'stock.location', 'Location', compute='_compute_package_info',
+        index=True, readonly=False, store=True)
+    company_id = fields.Many2one(
+        'res.company', 'Company', compute='_compute_package_info',
+        index=True, readonly=True, store=True)
+    owner_id = fields.Many2one(
+        'res.partner', 'Owner', compute='_compute_owner_id', search='_search_owner',
+        readonly=True, compute_sudo=True)
+    package_use = fields.Selection([
+        ('disposable', 'Disposable Box'),
+        ('reusable', 'Reusable Box'),
+        ], string='Package Use', default='disposable', required=True,
+        help="""Reusable boxes are used for batch picking and emptied afterwards to be reused. In the barcode application, scanning a reusable box will add the products in this box.
+        Disposable boxes aren't reused, when scanning a disposable box in the barcode application, the contained products are added to the transfer.""")
+    shipping_weight = fields.Float(string='Shipping Weight', help="Total weight of the package.")
+    valid_sscc = fields.Boolean('Package name is valid SSCC', compute='_compute_valid_sscc')
+    pack_date = fields.Date('Pack Date', default=fields.Date.today)
+
+    @api.depends('name', 'package_type_id.packaging_length', 'package_type_id.width', 'package_type_id.height')
+    @api.depends_context('formatted_display_name')
+    def _compute_display_name(self):
+        packages_to_process_ids = []
+        for package in self:
+            if package.env.context.get('formatted_display_name') and package.package_type_id and package.package_type_id.packaging_length and package.package_type_id.width and package.package_type_id.height:
+                package.display_name = f"{package.name}\t--{package.package_type_id.packaging_length} x {package.package_type_id.width} x {package.package_type_id.height}--"
+            else:
+                packages_to_process_ids.append(package.id)
+        if packages_to_process_ids:
+            super(StockPackage, self.env['stock.package'].browse(packages_to_process_ids))._compute_display_name()
+
+    @api.depends('quant_ids.location_id', 'quant_ids.company_id')
+    def _compute_package_info(self):
+        for package in self:
+            package.location_id = False
+            package.company_id = False
+            quants = package.quant_ids.filtered(lambda q: q.product_uom_id.compare(q.quantity, 0) > 0)
+            if quants:
+                package.location_id = quants[0].location_id
+                if all(q.company_id == quants[0].company_id for q in package.quant_ids):
+                    package.company_id = quants[0].company_id
+
+    @api.depends('quant_ids.owner_id')
+    def _compute_owner_id(self):
+        for package in self:
+            package.owner_id = False
+            if package.quant_ids and all(
+                q.owner_id == package.quant_ids[0].owner_id for q in package.quant_ids
+            ):
+                package.owner_id = package.quant_ids[0].owner_id
+
+    @api.depends('name')
+    def _compute_valid_sscc(self):
+        self.valid_sscc = False
+        for package in self:
+            if package.name:
+                package.valid_sscc = check_barcode_encoding(package.name, 'sscc')
+
+    def _search_owner(self, operator, value):
+        if operator in Domain.NEGATIVE_OPERATORS:
+            return NotImplemented
+        return Domain('quant_ids.owner_id', operator, value)
+
+    def write(self, vals):
+        if 'location_id' in vals:
+            if not vals['location_id']:
+                raise UserError(self.env._('Cannot remove the location of a non empty package'))
+            if any(not pack.quant_ids for pack in self):
+                raise UserError(self.env._('Cannot move an empty package'))
+            # create a move from the old location to new location
+            location_dest_id = self.env['stock.location'].browse(vals['location_id'])
+            quant_to_move = self.quant_ids.filtered(lambda q: q.quantity > 0)
+            quant_to_move.move_quants(location_dest_id, message=self.env._('Package manually relocated'))
+        return super().write(vals)
+
+    def unpack(self):
+        self.quant_ids.move_quants(message=self.env._("Quantities unpacked"), unpack=True)
+        # Quant clean-up, mostly to avoid multiple quants of the same product. For example, unpack
+        # 2 packages of 50, then reserve 100 => a quant of -50 is created at transfer validation.
+        self.quant_ids._quant_tasks()
+
+    def action_view_picking(self):
+        action = self.env["ir.actions.actions"]._for_xml_id("stock.action_picking_tree_all")
+        domain = ['|', ('result_package_id', 'in', self.ids), ('package_id', 'in', self.ids)]
+        pickings = self.env['stock.move.line'].search(domain).mapped('picking_id')
+        action['domain'] = [('id', 'in', pickings.ids)]
+        return action
+
+    def _check_move_lines_map_quant(self, move_lines):
+        """ This method checks that all product (quants) of self (package) are well present in the `move_line_ids`. """
+        precision_digits = self.env['decimal.precision'].precision_get('Product Unit of Measure')
+
+        def _keys_groupby(record):
+            return record.product_id, record.lot_id
+
+        if not move_lines:
+            return True
+
+        grouped_quants = {}
+        for k, g in groupby(self.quant_ids, key=_keys_groupby):
+            grouped_quants[k] = sum(self.env['stock.quant'].concat(*g).mapped('quantity'))
+
+        grouped_ops = {}
+        for k, g in groupby(move_lines, key=_keys_groupby):
+            grouped_ops[k] = sum(self.env['stock.move.line'].concat(*g).mapped('quantity'))
+
+        return all(float_is_zero(grouped_quants.get(key, 0) - grouped_ops.get(key, 0), precision_digits=precision_digits) for key in grouped_quants) \
+           and all(float_is_zero(grouped_ops.get(key, 0) - grouped_quants.get(key, 0), precision_digits=precision_digits) for key in grouped_ops)
+
+    def _get_weight(self, picking_id=False):
+        res = {}
+        if picking_id:
+            package_weights = defaultdict(float)
+            res_groups = self.env['stock.move.line']._read_group(
+                [('result_package_id', 'in', self.ids), ('product_id', '!=', False), ('picking_id', '=', picking_id)],
+                ['result_package_id', 'product_id', 'product_uom_id', 'quantity'],
+                ['__count'],
+            )
+            for result_package, product, product_uom, quantity, count in res_groups:
+                package_weights[result_package.id] += (
+                    count
+                    * product_uom._compute_quantity(quantity, product.uom_id)
+                    * product.weight
+                )
+        for package in self:
+            weight = package.package_type_id.base_weight or 0.0
+            if picking_id:
+                res[package] = weight + package_weights[package.id]
+            else:
+                for quant in package.quant_ids:
+                    weight += quant.quantity * quant.product_id.weight
+                res[package] = weight
+        return res

--- a/addons/stock/models/stock_package_level.py
+++ b/addons/stock/models/stock_package_level.py
@@ -13,7 +13,7 @@ class StockPackage_Level(models.Model):
     _check_company_auto = True
 
     package_id = fields.Many2one(
-        'stock.quant.package', 'Package', required=True, check_company=True,
+        'stock.package', 'Package', required=True, check_company=True,
         domain="[('location_id', 'child_of', parent.location_id), '|', ('company_id', '=', False), ('company_id', '=', company_id)]")
     picking_id = fields.Many2one('stock.picking', 'Picking', check_company=True, index='btree_not_null')
     move_ids = fields.One2many('stock.move', 'package_level_id')

--- a/addons/stock/models/stock_quant.py
+++ b/addons/stock/models/stock_quant.py
@@ -11,9 +11,7 @@ from psycopg2 import Error
 from odoo import _, api, fields, models
 from odoo.exceptions import UserError, ValidationError
 from odoo.fields import Domain
-from odoo.tools import SQL, groupby
-from odoo.tools.barcode import check_barcode_encoding
-from odoo.tools.float_utils import float_is_zero
+from odoo.tools import SQL
 
 _logger = logging.getLogger(__name__)
 
@@ -70,7 +68,7 @@ class StockQuant(models.Model):
     lot_properties = fields.Properties(related='lot_id.lot_properties', definition='product_id.lot_properties_definition', readonly=True)
     sn_duplicated = fields.Boolean(string="Duplicated Serial Number", compute='_compute_sn_duplicated', help="If the same SN is in another Quant")
     package_id = fields.Many2one(
-        'stock.quant.package', 'Package',
+        'stock.package', 'Package',
         domain="['|', ('location_id', '=', location_id), '&', ('location_id', '=', False), '&', ('package_use', '=', 'reusable'), ('quant_ids', '=', False)]",
         help='The package containing this quant', ondelete='restrict', check_company=True, index=True)
     owner_id = fields.Many2one(
@@ -276,7 +274,7 @@ class StockQuant(models.Model):
                 product = self.env['product.product'].browse(vals['product_id'])
                 location = self.env['stock.location'].browse(vals['location_id'])
                 lot_id = self.env['stock.lot'].browse(vals.get('lot_id'))
-                package_id = self.env['stock.quant.package'].browse(vals.get('package_id'))
+                package_id = self.env['stock.package'].browse(vals.get('package_id'))
                 owner_id = self.env['res.partner'].browse(vals.get('owner_id'))
                 quant = self.env['stock.quant']
                 if not self.env.context.get('import_file'):
@@ -1244,8 +1242,8 @@ class StockQuant(models.Model):
 
         :param location_id: `stock.location`
         :param location_dest_id: `stock.location`
-        :param package_id: `stock.quant.package`
-        :param package_dest_id: `stock.quant.package`
+        :param package_id: `stock.package`
+        :param package_dest_id: `stock.package`
         :return: dict with all values needed to create a new `stock.move` with its move line.
         """
         self.ensure_one()
@@ -1516,7 +1514,7 @@ class StockQuant(models.Model):
         """ Directly move a stock.quant to another location and/or package by creating a stock.move.
 
         :param location_dest_id: `stock.location` destination location for the quants
-        :param package_dest_id: `stock.quant.package´ destination package for the quants
+        :param package_dest_id: `stock.package` destination package for the quants
         :param message: String to fill the reference field on the generated stock.move
         :param unpack: set to True when needing to unpack the quant
         """
@@ -1534,150 +1532,3 @@ class StockQuant(models.Model):
                 result_package_id))
         moves = self.env['stock.move'].create(move_vals)
         moves._action_done()
-
-
-class StockQuantPackage(models.Model):
-    """ Packages containing quants and/or other packages """
-    _name = 'stock.quant.package'
-    _description = "Packages"
-    _order = 'name'
-
-    name = fields.Char(
-        'Package Reference', copy=False, index='trigram', required=True,
-        default=lambda self: self.env['ir.sequence'].next_by_code('stock.quant.package') or _('Unknown Pack'))
-    quant_ids = fields.One2many('stock.quant', 'package_id', 'Bulk Content', readonly=True,
-        domain=['|', ('quantity', '!=', 0), ('reserved_quantity', '!=', 0)])
-    package_type_id = fields.Many2one(
-        'stock.package.type', 'Package Type', index=True)
-    location_id = fields.Many2one(
-        'stock.location', 'Location', compute='_compute_package_info',
-        index=True, readonly=False, store=True)
-    company_id = fields.Many2one(
-        'res.company', 'Company', compute='_compute_package_info',
-        index=True, readonly=True, store=True)
-    owner_id = fields.Many2one(
-        'res.partner', 'Owner', compute='_compute_owner_id', search='_search_owner',
-        readonly=True, compute_sudo=True)
-    package_use = fields.Selection([
-        ('disposable', 'Disposable Box'),
-        ('reusable', 'Reusable Box'),
-        ], string='Package Use', default='disposable', required=True,
-        help="""Reusable boxes are used for batch picking and emptied afterwards to be reused. In the barcode application, scanning a reusable box will add the products in this box.
-        Disposable boxes aren't reused, when scanning a disposable box in the barcode application, the contained products are added to the transfer.""")
-    shipping_weight = fields.Float(string='Shipping Weight', help="Total weight of the package.")
-    valid_sscc = fields.Boolean('Package name is valid SSCC', compute='_compute_valid_sscc')
-    pack_date = fields.Date('Pack Date', default=fields.Date.today)
-
-    @api.depends('name', 'package_type_id.packaging_length', 'package_type_id.width', 'package_type_id.height')
-    @api.depends_context('formatted_display_name')
-    def _compute_display_name(self):
-        packages_to_process_ids = []
-        for package in self:
-            if package.env.context.get('formatted_display_name') and package.package_type_id and package.package_type_id.packaging_length and package.package_type_id.width and package.package_type_id.height:
-                package.display_name = f"{package.name}\t--{package.package_type_id.packaging_length} x {package.package_type_id.width} x {package.package_type_id.height}--"
-            else:
-                packages_to_process_ids.append(package.id)
-        if packages_to_process_ids:
-            super(StockQuantPackage, self.env['stock.quant.package'].browse(packages_to_process_ids))._compute_display_name()
-
-    @api.depends('quant_ids.location_id', 'quant_ids.company_id')
-    def _compute_package_info(self):
-        for package in self:
-            package.location_id = False
-            package.company_id = False
-            quants = package.quant_ids.filtered(lambda q: q.product_uom_id.compare(q.quantity, 0) > 0)
-            if quants:
-                package.location_id = quants[0].location_id
-                if all(q.company_id == quants[0].company_id for q in package.quant_ids):
-                    package.company_id = quants[0].company_id
-
-    @api.depends('quant_ids.owner_id')
-    def _compute_owner_id(self):
-        for package in self:
-            package.owner_id = False
-            if package.quant_ids and all(
-                q.owner_id == package.quant_ids[0].owner_id for q in package.quant_ids
-            ):
-                package.owner_id = package.quant_ids[0].owner_id
-
-    @api.depends('name')
-    def _compute_valid_sscc(self):
-        self.valid_sscc = False
-        for package in self:
-            if package.name:
-                package.valid_sscc = check_barcode_encoding(package.name, 'sscc')
-
-    def _search_owner(self, operator, value):
-        if operator in Domain.NEGATIVE_OPERATORS:
-            return NotImplemented
-        return Domain('quant_ids.owner_id', operator, value)
-
-    def write(self, vals):
-        if 'location_id' in vals:
-            if not vals['location_id']:
-                raise UserError(_('Cannot remove the location of a non empty package'))
-            if any(not pack.quant_ids for pack in self):
-                raise UserError(_('Cannot move an empty package'))
-            # create a move from the old location to new location
-            location_dest_id = self.env['stock.location'].browse(vals['location_id'])
-            quant_to_move = self.quant_ids.filtered(lambda q: q.quantity > 0)
-            quant_to_move.move_quants(location_dest_id, message=_('Package manually relocated'))
-        return super().write(vals)
-
-    def unpack(self):
-        self.quant_ids.move_quants(message=_("Quantities unpacked"), unpack=True)
-        # Quant clean-up, mostly to avoid multiple quants of the same product. For example, unpack
-        # 2 packages of 50, then reserve 100 => a quant of -50 is created at transfer validation.
-        self.quant_ids._quant_tasks()
-
-    def action_view_picking(self):
-        action = self.env["ir.actions.actions"]._for_xml_id("stock.action_picking_tree_all")
-        domain = ['|', ('result_package_id', 'in', self.ids), ('package_id', 'in', self.ids)]
-        pickings = self.env['stock.move.line'].search(domain).mapped('picking_id')
-        action['domain'] = [('id', 'in', pickings.ids)]
-        return action
-
-    def _check_move_lines_map_quant(self, move_lines):
-        """ This method checks that all product (quants) of self (package) are well present in the `move_line_ids`. """
-        precision_digits = self.env['decimal.precision'].precision_get('Product Unit')
-
-        def _keys_groupby(record):
-            return record.product_id, record.lot_id
-
-        grouped_quants = {}
-        for k, g in groupby(self.quant_ids, key=_keys_groupby):
-            grouped_quants[k] = sum(self.env['stock.quant'].concat(*g).mapped('quantity'))
-
-        grouped_ops = {}
-        for k, g in groupby(move_lines, key=_keys_groupby):
-            grouped_ops[k] = sum(self.env['stock.move.line'].concat(*g).mapped('quantity'))
-
-        if any(not float_is_zero(grouped_quants.get(key, 0) - grouped_ops.get(key, 0), precision_digits=precision_digits) for key in grouped_quants) \
-                or any(not float_is_zero(grouped_ops.get(key, 0) - grouped_quants.get(key, 0), precision_digits=precision_digits) for key in grouped_ops):
-            return False
-        return True
-
-    def _get_weight(self, picking_id=False):
-        res = {}
-        if picking_id:
-            package_weights = defaultdict(float)
-            res_groups = self.env['stock.move.line']._read_group(
-                [('result_package_id', 'in', self.ids), ('product_id', '!=', False), ('picking_id', '=', picking_id)],
-                ['result_package_id', 'product_id', 'product_uom_id', 'quantity'],
-                ['__count'],
-            )
-            for result_package, product, product_uom, quantity, count in res_groups:
-                package_weights[result_package.id] += (
-                    count
-                    * product_uom._compute_quantity(quantity, product.uom_id)
-                    * product.weight
-                )
-        for package in self:
-            weight = package.package_type_id.base_weight or 0.0
-            if picking_id:
-                res[package] = weight + package_weights[package.id]
-            else:
-                for quant in package.quant_ids:
-                    weight += quant.quantity * quant.product_id.weight
-                res[package] = weight
-        return res

--- a/addons/stock/models/stock_scrap.py
+++ b/addons/stock/models/stock_scrap.py
@@ -31,7 +31,7 @@ class StockScrap(models.Model):
         'stock.lot', 'Lot/Serial',
         domain="[('product_id', '=', product_id)]", check_company=True)
     package_id = fields.Many2one(
-        'stock.quant.package', 'Package',
+        'stock.package', 'Package',
         check_company=True)
     owner_id = fields.Many2one('res.partner', 'Owner', check_company=True)
     move_ids = fields.One2many('stock.move', 'scrap_id')
